### ISSUE TITLE
Very slow initial Library scan

### ISSUE THREAD

**Issue by justinrush** [jellyfin/jellyfin#2600]
**Describe the bug**
The initial scan of my library is taking a very long time - I'm at 24 hours and it is still going.  I initially added a small library (~400 titles with 2 copies of each) and it went pretty fast (30 minutes?).  Then I added my larger library (~2000 titles with 2 copies of each), and that has been going for 24 hours.

**System (please complete the following information):**
 - OS: Ubuntu 18.04
 - Virtualization: Docker (jellyfin/jellyfin) on Kubernetes 1.16.8
 - Clients: Android, Chrome
 - Browser: Chrome
 - Jellyfin Version: 10.5.0
 - Networking: NGNIX L7
 - Storage: Rook w/ m.2 SSDs

**To Reproduce**
1. Add a large library with multiple directories

**Expected behavior**
Multi-threaded discovery of the library with high CPU usage for 30 - 60 minutes.

**Logs**
For the most part, this:
```
[13:35:38] [INF] [19] App: MovieDbProvider: Finding id for item: The.Wedding.Ringer
[13:35:41] [INF] [12] App: MovieDbProvider: Finding id for item: The.Whole.Truth
[13:35:43] [INF] [12] App: MovieDbProvider: Finding id for item: The.Wo
[13:35:44] [INF] [12] App: MovieDbProvider: Finding id for item: The.Wolf.of.Wall.Street
[13:35:46] [INF] [19] App: MovieDbProvider: Finding id for item: The.World.According.to.Garp
[13:35:49] [INF] [99] App: MovieDbProvider: Finding id for item: The.Worlds.End
[13:35:51] [INF] [61] App: MovieDbProvider: Finding id for item: The.is.the.End
```

Although, sometimes this:
```
[13:34:18] [INF] [31] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Killing ffmpeg process
[13:34:18] [ERR] [31] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: ffmpeg image extraction failed for file:"/media/hd-2/The.Survivalist/The Survivalist.mkv"
[13:34:18] [ERR] [31] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: I-frame image extraction failed, will attempt standard way. Input: file:"/media/hd-2/The.Survivalist/The Survivalist.mkv"
System.Exception: ffmpeg image extraction failed for file:"/media/hd-2/The.Survivalist/The Survivalist.mkv"
at MediaBrowser.MediaEncoding.Encoder.MediaEncoder.ExtractImageInternal(String inputPath, String container, MediaStream videoStream, Nullable`1 imageStreamIndex, Nullable`1 threedFormat, Nullable`1 offset, Boolean useIFrame, CancellationToken cancellationToken)
at MediaBrowser.MediaEncoding.Encoder.MediaEncoder.ExtractImage(String[] inputFiles, String container, MediaStream videoStream, Nullable`1 imageStreamIndex, MediaProtocol protocol, Boolean isAudio, Nullable`1 threedFormat, Nullable`1 offset, CancellationToken cancellationToken)
[13:34:23] [ERR] [65] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: ffmpeg image extraction failed for file:"/media/hd-2/The.Survivalist/The Survivalist.mkv"
[13:34:23] [ERR] [65] App: Error in Screen Grabber
System.Exception: ffmpeg image extraction failed for file:"/media/hd-2/The.Survivalist/The Survivalist.mkv"
at MediaBrowser.MediaEncoding.Encoder.MediaEncoder.ExtractImageInternal(String inputPath, String container, MediaStream videoStream, Nullable`1 imageStreamIndex, Nullable`1 threedFormat, Nullable`1 offset, Boolean useIFrame, CancellationToken cancellationToken)
at MediaBrowser.MediaEncoding.Encoder.MediaEncoder.ExtractImage(String[] inputFiles, String container, MediaStream videoStream, Nullable`1 imageStreamIndex, MediaProtocol protocol, Boolean isAudio, Nullable`1 threedFormat, Nullable`1 offset, CancellationToken cancellationToken)
at MediaBrowser.Providers.MediaInfo.VideoImageProvider.GetVideoImage(Video item, CancellationToken cancellationToken)
at MediaBrowser.Providers.Manager.ItemImageProvider.RefreshFromProvider(BaseItem item, IDynamicImageProvider provider, ImageRefreshOptions refreshOptions, TypeOptions savedOptions, LibraryOptions libraryOptions, ICollection`1 downloadedImages, RefreshResult result, CancellationToken cancellationToken)
```
and sometimes this:
```
[13:36:55] [INF] [89] App: MovieDbProvider: Finding id for item: Trash
[matroska,webm @ 0x562da606a600] Could not find codec parameters for stream 2 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x562da606a600] Could not find codec parameters for stream 3 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x562da606a600] Could not find codec parameters for stream 4 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x562da606a600] Could not find codec parameters for stream 5 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x562da606a600] Could not find codec parameters for stream 6 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x5595840e9600] Could not find codec parameters for stream 2 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x5595840e9600] Could not find codec parameters for stream 3 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x5595840e9600] Could not find codec parameters for stream 4 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x5595840e9600] Could not find codec parameters for stream 5 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x5595840e9600] Could not find codec parameters for stream 6 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size
Consider increasing the value for the 'analyzeduration' and 'probesize' options
```

**Additional context**
The process is doing a few hundred write IOs with around 8ms latency.  
Hardly any CPU and maybe a GB of memory is currently consumed by it.
Looks like /config is about 640MB used so far

Also I initially ran into a problem with too many open file watches.  I upped that in kernel params and bounced the container and it went away.  I notice that you can disable watch files in advanced settings after I added the library - you may want to consider calling out that folks need to change this param if they keep the default setting (8096?):
fs.inotify.max_user_watches=1048576

Hardware on the server is Intel(R) Core(TM) i7-8559U CPU @ 2.70GHz

**Comment by Artiume** [comment]
> Also I initially ran into a problem with too many open file watches. I upped that in kernel params and bounced the container and it went away. I notice that you can disable watch files in advanced settings after I added the library - you may want to consider calling out that folks need to change this param if they keep the default setting (8096?):
fs.inotify.max_user_watches=1048576

That is located here, but yeah it's sparse. I'll add instructions on how to increase it. https://jellyfin.org/docs/general/server/libraries.html#real-time-monitoring 

> Consider increasing the value for the 'analyzeduration' and 'probesize' options
[matroska,webm @ 0x562da606a600] Could not find codec parameters for stream 3 (Subtitle: hdmv_pgs_subtitle (pgssub)): unspecified size

This is probably fixed by https://github.com/jellyfin/jellyfin/pull/2582 

> [13:34:18] [INF] [31] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Killing ffmpeg process [13:34:18] [ERR] [31] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: ffmpeg image extraction failed for file:"/media/hd-2/The.Survivalist/The Survivalist.mkv" [13:34:18] [ERR] [31] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: I-frame image extraction failed, will attempt standard way. 

This is the Screen Grabber failing, tracked here, i believe it's docker related issue so probably a package missing https://github.com/jellyfin/jellyfin/issues/2355

hopefully we can fix screen grabber soon but everything should be fixed in 10.5.1 :)

**Comment by justinrush** [comment]
@Artiume thats great!  what about something that is multi-threaded?  I suspect my initial scan will still take a long time, but I've got plenty of CPU if you can multi-thread it.

**Comment by Artiume** [comment]
Overall, the scan system needs to be reworked. First scan all the files and put them in the dB so that they're accessible. Then you do the stage of metadata grabbing. I think multi threading is do able. The two processes being independent will go a long way to increase efficiency. 


Found this
https://github.com/jellyfin/jellyfin/blob/5c669d7ad7145eb2b54ec88a8636a28ea4eefcd1/MediaBrowser.Api/Library/LibraryService.cs#L409

**Comment by evulhotdog** [comment]
We ain't closin' this one. Its still happening.

**Comment by sergiotapia-papa** [comment]
Moving from plex to jellyfin, and I have a 4000+ movie collection, this is going to take a while... it's been 30 minutes and so far it's only detected 213 movies. Appreciate any help on how to speed this up. It's instant on Emby and Plex.

**Comment by sergiotapia-papa** [comment]
645 scanned now, so ~400 movies per 30 minutes rate.

**Comment by ranaldsgift** [comment]
It's not only the initial scan apparently. Here's my latest scan:

![Latest scan.](https://i.imgur.com/CgKzFCf.png)

**Comment by justinrush** [comment]
My scans never complete, i gave up for a while until this is fixed and continue using plex

**Comment by ranaldsgift** [comment]
@justinrush Just curious, do you have missing items enabled on your library? I do and I'm wondering if it could be related to that.

**Comment by ssyangchn** [comment]
7k+ movies and 2000+ tv take me about 2weeks to scan. Still 400+ tv can't be find even I have rescan a lot of times o(╥﹏╥)o

**Comment by ranaldsgift** [comment]
@ssyangchn Yikes! What a disaster. Hope this issue gets resolved for the next major release!

**Comment by bugfixin** [comment]
@ssyangchn Mind giving us a pastebin of a full directory listing of your library to test against as well as some logs? Also have you checked to make sure you don't have chapter image extraction or similar intensive options on?

**Comment by evulhotdog** [comment]
> @ssyangchn Mind giving us a pastebin of a full directory listing of your library to test against as well as some logs? Also have you checked to make sure you don't have chapter image extraction or similar intensive options on?

I had a feeling like it was maybe related to this @bugfixin https://github.com/jellyfin/jellyfin/issues/2355

**Comment by ranaldsgift** [comment]
Just to chime in, my library has chapter image extraction disabled and still takes extremely long. 5k movies + 100k episodes, took over 18 hours last time I ran the scan - and realtime monitoring is seemingly broken, so I can't simply use that instead.

**Comment by dkanada** [comment]
In general, we realize that scanning is extremely slow right now, mainly because it's the one section of the codebase almost no one has touched since the fork. We have been waiting on the API and database migrations because those will fundamentally change a large portion of the codebase and needed to happen before any scanning improvements.

The API migration was finished recently and we are currently tracking down any regressions, and the database migration is still in progress, but that's the best estimate I can really give for substantial speed increases for scans. Even once those are finished, we'll need to wait for someone who has a lot of free time and an interest in untangling a mess of code to fix the scanning, but the goal is around 10.8.0 or 10.9.0 if we can manage it.

**Comment by ranaldsgift** [comment]
Really appreciate the response, and definitely understandable considering the current mess of the code. For me personally, the extreme scan time isn't terrible if it's rarely ever needed, and this was always the case for me - until now. I have an issue as mentioned in #3847 that has a major impact on realtime monitoring picking up new items. I have library scans disabled, with realtime monitoring on, and I will see in the logs mentions about finding new items, but they don't appear in the library. If I look at my libraries in the Dashboard, I will see a percentage indicator on one of them as if it is scanning. Eventually the log floods with OnRefreshStart/OnRefreshComplete and items in the library are only updated at the end of this "scan" which lasts as long as a normal scan so upwards of 18 hours.

The only way to reliably populate my libraries is to simply run the manual scan every few days. My new items are delayed, but at least I'm not scanning my media 24/7.

**Comment by dennis800121** [comment]
...Same her. I have 16 000 movies and it takes forever!

**Comment by deyarmond** [comment]
I just migrated from Emby, jellyfin has been doing the initial scan for over 2 hours now and has only picked up 21 movies so far. Still another 300+ to go and 1000's of episodes so it'll probably be a week or so before it finishes the scan at this rate. I don't recall Emby taking this long to scan in.

**Comment by deyarmond** [comment]
It's been almost 24 hours continuous scanning for me and it hasn't even started the TV shows yet.

**Comment by Artiume** [comment]
I need you to look into where the issue is. I've been dealing with this issue for over a year and I assure you that it isn't THAT bad. I use rclone and gdrive and it takes me 2 to 3 days to scan 100+ TB of content. There's something else going on with your system. I get that you're frustrated but you need to investigate more

**Comment by deyarmond** [comment]
Sure thing. Again I don't want to sound complainy and I'll work to find a solution. First things first, where should I begin to look? I stopped the scanning and purged jellyfin from my system so I will try a clean reinstall and see where the library scan is in an hour or so. I recall Emby taking around 3-4 hours to do a 3TB scan.

It hung around 87.7% yesterday when I posted my first message and when I got home from work today on my 2nd message it got up to 88.1%

**Comment by ranaldsgift** [comment]
Normally I would agree, but I think we need a bit more information about their Emby install. Were there any issues with the scan on Emby? Are Emby and Jellyfin using the same system, same library file locations, etc? If Emby runs the scan fine and it is using the same configuration as Jellyfin would that not indicate it may be a JF issue? Not with certainty, but it seems like it might lean that way.

**Comment by Artiume** [comment]
First things first, the scanner percentage is broken AF, ignore it and watch the logs. Does it happen with only a movie dB or only the TV show dB. Perhaps it's stuck on music. Add each type one at a time and see if there's a specific category that is having issues.

**Comment by deyarmond** [comment]
Sep 03 16:13:35 osmc jellyfin[1132]: [16:13:35] [WRN] HTTP Response 200 to 192.168.2.60. Time (slow): 0:00:04.3572444. http://192.168.2.66:8096/Users/05fd593a52304689b9
Sep 03 16:13:35 osmc jellyfin[1132]: [16:13:35] [WRN] HTTP Response 200 to 192.168.2.60. Time (slow): 0:00:01.8878441. http://192.168.2.66:8096/Users/05fd593a52304689b9
Sep 03 16:13:35 osmc jellyfin[1132]: [16:13:35] [WRN] HTTP Response 200 to 192.168.2.60. Time (slow): 0:00:01.9922715. http://192.168.2.66:8096/Users/05fd593a52304689b9
Sep 03 16:13:45 osmc jellyfin[1132]: [16:13:45] [WRN] HTTP Response 200 to 192.168.2.60. Time (slow): 0:00:01.188069. http://192.168.2.66:8096/System/Info
Sep 03 16:13:45 osmc jellyfin[1132]: [16:13:45] [WRN] HTTP Response 200 to 192.168.2.60. Time (slow): 0:00:01.2427977. http://192.168.2.66:8096/ScheduledTasks
Sep 03 16:13:45 osmc jellyfin[1132]: [16:13:45] [WRN] HTTP Response 200 to 192.168.2.60. Time (slow): 0:00:01.3222296. http://192.168.2.66:8096/Sessions?ActiveWithinSec
Sep 03 16:13:45 osmc jellyfin[1132]: [16:13:45] [WRN] HTTP Response 200 to 192.168.2.60. Time (slow): 0:00:01.3070208. http://192.168.2.66:8096/System/ActivityLog/Entri
Sep 03 16:13:45 osmc jellyfin[1132]: [16:13:45] [WRN] HTTP Response 200 to 192.168.2.60. Time (slow): 0:00:01.3117551. http://192.168.2.66:8096/System/ActivityLog/Entri
Sep 03 16:13:45 osmc jellyfin[1132]: [16:13:45] [WRN] HTTP Response 200 to 192.168.2.60. Time (slow): 0:00:01.4360744. http://192.168.2.66:8096/LiveTv/Recordings?UserId

This is pretty much all that's in the log files. I don't have a music library or live tv just movies and TV shows.

**Comment by Artiume** [comment]
Those are all web responses, they're independent of the scan. Scan movies and TV shows seperately.

**Comment by deyarmond** [comment]
My mistake, I guess I didn't run sudo autoremove. I'm giving that a go, didn't seem to clear out jellyfin completely. I'll take a look at the log once I start a single scan and see what we got.

**Comment by deyarmond** [comment]
Since an hour ago when I started my scan of just TV shows, it has added I think 9 shows and scanned through most others but hasn't added anything. Log attached.

https://pastebin.com/wnTxBuJ3

**Comment by deyarmond** [comment]
So I finally got my media scanned in after letting it run overnight and all morning. Now the issue lies with letting the jellyfin plugin for Kodi do a scan as it seems just as slow to scan media to the Kodi library as compared to Emby. Been running a scan about 20 minutes and it has scanned 3 TV shows to the Kodi library. Compared to Emby, that only took about 5 minutes to scan my entire library.

I might be way off as I'm not a programmer, just a tinkerer. But I feel like there is something seriously wrong with the way jellyfin scans its media. If there's any way of testing I would be happy to try and pinpoint where the issue is as this behaviour only appears in jellyfin.

**Comment by cvium** [comment]
Is this still an issue in 10.7?

**Comment by evulhotdog** [comment]
Honestly, haven't used Jellyfin because of this issue, but https://github.com/jellyfin/jellyfin/issues/2355 appears to be open, so I'm guessing it's still prevalent.

**Comment by douglasparker** [comment]
I've installed Jellyfin for the first time alongside Plex and specifically followed this issue because of the length of time the initial scan was taking, so I would agree and say that it is still an issue.

**Comment by orzel** [comment]
Oh, yes it is. Movies keep on appearing at a very very slow rate here. It has started days (weeks?) ago.  There are "only" ~4000 movies. Almost no resources is used (cpu, network, disk i/o), so i initially thought jellyfin was done and didn't manage to do anything.

**Comment by Artiume** [comment]
what version are you running and what settings do you have in your library such as chapter extraction

**Comment by LordHenko** [comment]
I wanna use Jellyfin so badly but the library scans are just way too slow, I hope this gets fixed soon.

**Comment by vbaros** [comment]
I did not have this problem since 10.7.5 version. Using on Synology in docker.

**Comment by scubbo** [comment]
Chiming in here to say that I have similar problems. 1.2Tb library - scan percentage got up to ~50% within a couple of minutes and then slowed to a crawl, ~18 hours later it's still at 84.1%. RaspberryPi4 with latest Docker image. I'll be happy to upload whatever logs you want, or experiment with partial scans, if you tell me what would be helpful!

**Comment by Pecon** [comment]
Just set up Jellyfin 10.7.6 on Raspbian stretch (no docker), raspberry pi 3.  Can confirm this is still an issue, currently taking multiple days to scan a library of ~300 files. Settings are practically default, media files are on an external hard drive. It is making progress, it is just painfully slow.

**Comment by jameskimmel** [comment]
I have the same problem. 
Ubuntu with mounted nfs share.
No VM or Docker.

I get that the initial scan can be slow. But why does it take 11 hours for every rescan?

**Comment by Pecon** [comment]
I didn't end up using Jellyfin for unrelated reasons, but I figured mentioning this might help someone else. When I had this issue, I noticed that disabling Jellyfin's 'watch for changes' feature in the library settings seemingly cleared up this horribly slow scan. It could just be coincidence that it worked for me, I didn't attempt to recreate the issue and try from fresh.

**Comment by jameskimmel** [comment]
You are right! Scan is now down from 11 hours to 8min. 
Besides, watch for changes did not work in my case anyway because NFS is not supported.
Thank you very much!

**Comment by yslupdates** [comment]
Any potential workarounds for this? I'm on 10.7.7 via Dietpi on a Pi 4 1GB but my very first scan brings my Pi to a slow crawl. Load averages in mid 10s most of the time during the scan.

**Comment by Pecon** [comment]
@yslupdates Did you try disabling the 'watch for changes' setting for the library you're scanning?

**Comment by yslupdates** [comment]
> @yslupdates Did you try disabling the 'watch for changes' setting for the library you're scanning?

Yes already have tried that. I even tried adding only 1 tv series and no other videos and that still slows my Pi 4 to a crawl. Default settings seem to be YMMV for debian when scanning I guess. I've seen some success stories but can't figure out what I'm doing wrong.

**Comment by partoneoftwo** [comment]
Ok. This issue has been up for 1.5 years. This fucking sucks. I hoped to migrate onto a better platform than Plex. If it cant import my movie library, which the solution is made for then I have to keep looking. With due respects to the efforts that have been done do build this, I don't  understand how you have managed to get here where a large library cant be imported within a sane timeframe. Very few computing jobs should take more than 24 hours to complete in 2021.

**Comment by cvium** [comment]
And a merry christmas to you too

**Comment by yslupdates** [comment]
> > @yslupdates Did you try disabling the 'watch for changes' setting for the library you're scanning?
> 
> Yes already have tried that. I even tried adding only 1 tv series and no other videos and that still slows my Pi 4 to a crawl. Default settings seem to be YMMV for debian when scanning I guess. I've seen some success stories but can't figure out what I'm doing wrong.

I ended up moving to a Pi 4 4gb up from a 1gb and all my problems are gone. I use Dietpi and initial scan takes maybe 2 hours but I can use other apps while it’s running now, probably due to more ram.

**Comment by evulhotdog** [comment]
I gave up on Jellyfin when I came upon this issue, but I’m using an i7 3700k with 32gb of ram when I experienced it.

**Comment by anthonylavado** [comment]
Except for the first post, I don't think anyone else who has experienced this issue has mentioned what their file structure is.

Another team member did ask someone earlier in this thread (see https://github.com/jellyfin/jellyfin/issues/2600#issuecomment-676704455) but we didn't get an answer there either.

So to those currently seeing this - how are the files organized in the problem libraries? Is everything in sub folders? Is everything just in one large folder?

**Comment by evulhotdog** [comment]
It’s the default radarr format of:

Movie Name (year)/movie.name.mkv

**Comment by cvium** [comment]
The reason it hasn't been fixed is because we can't reproduce it in our own setups.

If someone can create a minimal setup with dummy files or similar we can easily fix most things. Just look at #6906

**Comment by partoneoftwo** [comment]
My structure is:
**MOVIES** 
MOVIE TITLE (YEAR QUALITY) \ MOVIE FILE

**SHOWS** 
SHOW TITLE \ SEASON NUMBER \ MOVIE FILE

CPU is Dual 24 thread Xeon V2 2697, 100 gb RAM, non busy drives

I have read a suggestion to disable watch for changes.

**Comment by MikeMatau** [comment]
Got the same issue... My filestructure is:

Movies:
/Movies/Movie A (YYYY)/Movie A (YYYY) - [1080p DTS x265].mkv
/Movies/Movie A (YYYY)/Movie A (YYYY) - [1080p AC3 x264].mkv
/Movies/Movie B (YYYY)/Movie B (YYYY) - [1080p DTS x265].mkv
/Movies/Movie B (YYYY)/Movie B (YYYY) - [1080p AC3 HEVC].mkv
etc.

Maybe a session together using teamviewer or similar would help? At least I could offer that..

**Comment by Aokromes** [comment]
i confirm the 
[13:34:18] [INF] [31] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Killing ffmpeg process

part.

```spoiler
[08:32:29] [INF] [366] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: Killing ffmpeg process
[08:32:29] [ERR] [366] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: ffmpeg image extraction failed for file:"E:\T\The Blacklist\Temporada 5\5x16 - The Blacklist [x265].mkv"
[08:32:29] [ERR] [366] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: I-frame or HDR image extraction failed, will attempt with I-frame extraction disabled. Input: file:"E:\T\The \Temporada 5\5x16 - The [x265].mkv"
System.Exception: ffmpeg image extraction failed for file:"E:\T\The \Temporada 5\5x16 - The [x265].mkv"
   at MediaBrowser.MediaEncoding.Encoder.MediaEncoder.ExtractImageInternal(String inputPath, String container, MediaStream videoStream, Nullable`1 imageStreamIndex, Nullable`1 threedFormat, Nullable`1 offset, Boolean useIFrame, Boolean allowTonemap, CancellationToken cancellationToken)
   at MediaBrowser.MediaEncoding.Encoder.MediaEncoder.ExtractImage(String inputFile, String container, MediaStream videoStream, Nullable`1 imageStreamIndex, MediaSourceInfo mediaSource, Boolean isAudio, Nullable`1 threedFormat, Nullable`1 offset, CancellationToken cancellationToken)
[08:32:32] [ERR] [366] Jellyfin.Plugin.Tvdb.Providers.TvdbEpisodeImageProvider: Episode nullxnull not found for series 319073
[08:32:33] [ERR] [340] Jellyfin.Plugin.Tvdb.Providers.TvdbEpisodeProvider: Failed to retrieve episode with id null, series id null
TvDbSharper.TvDbServerException: The given series ID does not exist; No results for your query: map[AiredSeason:1 EpisodeNumber:21]
   at TvDbSharper.Infrastructure.Parser.Parse[T](ApiResponse response, IReadOnlyDictionary`2 errorMap)
   at TvDbSharper.Clients.SeriesClient.GetEpisodesAsync(Int32 seriesId, Int32 page, EpisodeQuery query, CancellationToken cancellationToken)
   at Jellyfin.Plugin.Tvdb.TvdbClientManager.<>c__DisplayClass24_0`1.<<TryGetValue>b__0>d.MoveNext()
```

**Comment by J4gQBqqR** [comment]
Same problem here. I have my library sorted, renamed using the servarr tools. If Jellyfin has a specific requirement of folder structure to make it "blazing fast", could you, as experienced user and know-how developer, provide us a detailed instruction on how to organize our folders? For us as end users, we simply need a good document to guide us through the jungle. I would argue that a bash script or a Github issue is not a good ducumentation. @anthonylavado

**Comment by davidtakac** [comment]
I'd just like to confirm that disabling "Real time monitoring" fixes the infinite load issue for me as well.

**Comment by Bhooty** [comment]
> I'd just like to confirm that disabling "Real time monitoring" fixes the infinite load issue for me as well.

Thanks.. Atlast got it to work thanks to you. If real time monitoring is not turned off the server don't take speed

**Comment by exaveal** [comment]
I'm new to Jellyfin and installed version 10.7.7. Jellyfin everything in default settings.
Library Scan took 2 days for 917 Movies, 7000 Songs, 2125 Episodes of Series.
When adding an album (max. 20 Songs) to the music library, Jellyfin cannot be used for around 30 minutes. The player just shows the loading circle.

That's really frustrating.

**Comment by Shadowghost** [comment]
Are you using local or cloud storage? The latter makes scanning really slow especially if you use NFOs and posters.

**Comment by exaveal** [comment]
Local storage and no nfo files and/or posters.

The bigger problem is when adding single albums / movies or series afterwards. E.g. when I add a music album. JellyFin Webinterface is incredible slow so it can't be used for about 30 minutes.

**Comment by davidtakac** [comment]
> I'd just like to confirm that disabling "Real time monitoring" fixes the infinite load issue for me as well.

This was actually not the fix for my exact issue. I tried to stream my media with the web interface and it was really slow for some videos, with load times of 20 mins or more, and I falsely attributed it to library scans. Turning off real-time monitoring seemed to only help a little bit. 

The real issue is that I didn't turn on server transcoding and the web interface doesn't support streaming most codecs with this off. In the end I tried the JellyCon addon for Kodi which puts the decoding load on the client. Jellyfin is only responsible for generating metadata and serving it to Kodi which handles decoding and everything works as expected.

**TL;DR: Slow web interface may be due to the fact that you're not using server transcoding and trying to stream media from the browser. Either turn on transcoding or use the JellyCon addon for Kodi to put decoding load on the client.**

**Comment by psiberfunk** [comment]
Because there seems to be a lack of multi-threading, the web interface , and really any use of Jellyfin by a client like Kodi or Swiftfin, for example, is totally unusable during the initial indexing.  Hangs/timeouts and/or extremely slow load times are pretty standard.  At a minimum, it would be ideal to fork out the indexing code into it's own process if it runs longer than 10 seconds or something.  Having it block basically every other part of Jellyfin and causing it to become completely unresponsive is no bueno.

**Comment by cvium** [comment]
> Because there seems to be a lack of multi-threading, the web interface , and really any use of Jellyfin by a client like Kodi or Swiftfin, for example, is totally unusable during the initial indexing.  Hangs/timeouts and/or extremely slow load times are pretty standard.  At a minimum, it would be ideal to fork out the indexing code into it's own process if it runs longer than 10 seconds or something.  Having it block basically every other part of Jellyfin and causing it to become completely unresponsive is no bueno.

Ignorant statement. It runs in its own thread pool thread.

**Comment by psiberfunk** [comment]
> Ignorant statement. It runs in its own thread pool thread.

I apologize, I was basing that threading assumption based on what someone had said about the code not being changed much since the fork from Emby.  If it really runs in it's own threadpool, why does the whole Jellyfin server slow to a crawl during this process?  I get tons of slow web response warnings and 30-second page load times and the like during the the indexing.. it would seem that SOMETHING is blocking the other parts of Jellyfin from responding... so it sure doesn't feel like its multithreaded.

**Comment by cvium** [comment]
The thing about that thread pool is that it's shared in the whole application, so whenever a web request comes in, it's handled on a thread pool thread. Thread pool starvation can be an issue. But the biggest problem is likely the database connection. The database is frequently read and written to during scanning, which will impact loading data for the frontends.

**Comment by psiberfunk** [comment]
> The thing about that thread pool is that it's shared in the whole application, so whenever a web request comes in, it's handled on a thread pool thread. Thread pool starvation can be an issue. But the biggest problem is likely the database connection. The database is frequently read and written to during scanning, which will impact loading data for the frontends.

Interesting.. I wonder what Emby is doing differently then to remain responsive in this situation ?  I tested them side by side and got very different experiences.

**Comment by xenago** [comment]
It's shockingly slow, and while it is running the entire UI is unusable. Wtf?

**Comment by ecwilsonaz** [comment]
On a Synology DS1812+, I'm noticing that during the initial scan (now running 2 days), the CPU/RAM usage is low, but in addition to causing the Jellyfin UI to grind to a halt, it's also affecting my server in other ways. The Synology UI is very slow. Even SSH is painfully slow. Here are some charts for the overall server (not just the JF docker instance):

<img width="766" alt="image" src="https://user-images.githubusercontent.com/2790907/210031736-6f8e02bf-e109-4d15-80d1-4d25f2d4499a.png">
<img width="773" alt="image" src="https://user-images.githubusercontent.com/2790907/210031319-fca83841-aba1-4ed6-8f0b-36c14ea72233.png">
<img width="774" alt="image" src="https://user-images.githubusercontent.com/2790907/210031483-cded9fcf-f064-4874-8f90-fda1c8bda5b8.png">
<img width="767" alt="image" src="https://user-images.githubusercontent.com/2790907/210031642-21dbb3c7-3bca-4860-9c00-595c96a9661f.png">
<img width="781" alt="image" src="https://user-images.githubusercontent.com/2790907/210031682-99e56c34-193e-471c-bae6-85ba0e61b264.png">

The I/O wait stands out. I'm surprised at all the I/O. Does the scan really have to _read_ each media file vs. just parsing file names?

**Comment by mantlabs** [comment]
I am having the same issue with 10.8.9 version. Any updates on this? The whole UI hangs and the scanning takes at least 3 full days to complete... This issue is THE biggest issue of Jellyfin in my opinion and should be addressed soon since it makes user experience very bad. A scan taking days... It moves people away from Jellyfin.

Is there any workaround to try? Or any ETA of when this will be fixed? Is it planned in a future version?

**Comment by nealey** [comment]
I did a cursory look into the source code for causes of this. What I found, which has already been mentioned previously in this issue, is that it's most likely one of two things:

1. Thread exhaustion by recursively spinning up too many asynchronous tasks: instead of queuing tasks for a pool of workers, the code appears to just kick up a bunch of async Tasks. The web server also uses this Task system (I assume) and the task to service an HTTP request has to stand in line, essentially.
2. Database contention: during scans, a ton of database modifications need to happen, and anything like a lock around data about to change could really slow down reads

My personal opinion is that it's the first one. Something I'd like to test, but haven't had time yet, is whether scanning a nested directory structure triggers the slowdown more consistently. My hypothesis is that a flat directory structure (all your files at the top level) would not cause any noticeable slowdowns, whereas a nested directory structure (lots of subdirectories, maybe with subdirectories in them) would trigger this bug.

Please note I'm just an interloper, not a core Jellyfin developer, or even a code contributor. I'm just a Go programmer who spent a couple hours poking through the source code with this issue in mind. Everything I wrote here may be completely wrong.

**Comment by mantlabs** [comment]
Folder structure:

Music Artist 1/
    Music Album 1/
        Song 1
        Song 2
        etc
    etc
etc

Having the above structure and also using the docker version if that helps to reproduce it.

Could I provide anything else in order to help reproduce it?

**Comment by mantlabs** [comment]
> 1. the code appears to start up a whole bunch of async Tasks. The web server also uses this Task system (I assume, haven't verified yet) and the task to service an HTTP request has to stand in line, essentially.

Could you please share the file location of the committed code you saw this in order for me to check too?

**Comment by nealey** [comment]
> Could I provide anything else in order to help reproduce it?

If you can now put the same files into a flat directory, and test indexing just that flat directory, that would at least confirm or disprove my hypothesis that directory nesting is what causes the problem. You don't have to destroy your existing structure, and you actually don't even need to do a copy. If you're using Unix, the following might work:

```sh
mkdir /path/to/test/directory
cd /path/to/test/directory
find /path/to/actual/music/library -type f -exec ln {} . \;
```

This should make a hard link for each file, which is essentially telling the filesystem to make a new file that uses the same data. It will be safe to remove these files after testing, but modifying them will modify the original, too.

Next, start up a new Jellyfin and have it scan this new library with no subdirectories. If the UI is still slow, then I was wrong, and the problem is something else. If the UI is more responsive, then we have something the developers can work with!

I don't remember exactly where I was looking, but [PerformLibraryValidation](https://github.com/jellyfin/jellyfin/blob/master/Emby.Server.Implementations/Library/LibraryManager.cs#L1052]PerformLibraryValidation) might be a place to start.

**Comment by wolk5** [comment]
I have the same issue. I have a few thousand music files (not flat directory) and it takes ages on the synology nas (dockered jellyfin) to do the initial scan. During this scan, the docker portal shows 100 % cpu usage although the containers in total are not at 100 % usage. It seems like there are constant spikes which are not being visualized.
![Screenshot 2023-02-22 144328](https://user-images.githubusercontent.com/85890593/220638507-248ff87d-d4cb-40d9-a528-6af36ee228a0.png)

**Comment by nealey** [comment]
> I went ahead and instrumented this test, making a completely flat directory structure of hard links to my directory-nested tree media. It seems promising.
> 
> I don't have any numbers for this, just feel. I'll let it go for a few days, keep using it, and then try the same thing with nested directories, to compare.

After a few days of testing, I no longer feel like I understand how to replicate the slowdowns I'm experiencing. I'm sure they're happening, I just can't figure out what triggers them. It does not seem to be strongly correlated to directory structure, though.

If anybody can reliably trigger the slowdowns, and, crucially, *figure out a workaround*, that's probably the push the development team needs to really investigate what's happening. They need some way to reproduce it in testing, to tell whether their work is improving the problem.

**Comment by 0xKopyKat** [comment]
Interesting discovery today. I’ve been testing out a Jellyfin side by side with Plex in Docker containers on my Synology NAS and been experiencing the dreaded library sync that takes days to finish, on a moderately small library (1000 movies).

While troubleshooting many of the symptoms in this issue, I noticed that Plex would have a burst of IO activity (observed in iotop) as soon as I spun up the Jellyfin container. Curious, I stopped the Plex container and to my elated disbelief, saw the IO wait drop from mid 40s down to 7ish. A few minutes later I noticed the jellyfin library scan, that had taken 2 days to add 200ish movies, had completed successfully. Giddy, I added in another library and it was done in a matter of minutes.

Could it be that it’s some kind of file lock contention that’s causing the scanning threads to deadlock and then smash the disk with retries? I was seeing large IOPS during the scan, which was bringing everything on the array to a grinding halt.

Looking back through this issue, there seems to be a fair number of people trying out Jellyfin side by side with Plex, and I’m wondering if that’s why it’s been difficult to reproduce?

If anyone is in the same boat as me, it may be worth stopping Plex (risking the wrath of your users) and seeing if anything improves for your Jellyfin Library scans.

**Comment by Pecon** [comment]
When I had this issue I was not running anything else of note on the raspberry pi I ran it on. Instead I discovered that disabling the filesystem watching was what fixed it for me. Since then I've seen a few other people in this thread discover that as their solution, and a number of others stating that didn't fix it for them. There are probably a few different situations/setups that are able to reproduce this issue, unfortunately.

**Comment by bigverm23** [comment]
this is still an issue.  Just spun up an instance through docker, and my small 1000 file library is at over 30 mins thus far to scan in.

I do have an instance of emby running as well...

**Comment by zachfi** [comment]
I've got the library mounted over NFS and so I expected it to be slower, but days later it was still not finished.  Disabling "Real-time monitoring" on the library allowed the scan to complete in minutes.  This should have been obvious since the underlying filesystem won't support it.

**Comment by NostalgicGnome** [comment]
I recently had to completely rebuild my database, and it has taken several days (it's still not entirely complete, but the initial discovery without plugins completed after ~2 days. I am just waiting on a re-scan now that I've re-enabled plugins.)

I use ZFS on the server on which I have all my media files (running Jellyfin in a docker container), and while I was waiting for the library scan to complete I found [this write up](https://github.com/teambvd/UnRAID-Performance-Compendium/blob/main/containers/sonarrRadarrLidarr.md) on some optimizations someone tried on a similar setup to mine. That write up just mentions the *arr apps, but I figured since Jellyfin also uses sqlite why not give it a shot! To my delight, applying (**important**: _almost_, not exactly, see below) those same steps seems to have _drastically_ improved start up times, library scan times, and basically everything that relies on operations on `library.db`.

Slightly adapting what [teambvd](https://github.com/teambvd) wrote in his post, essentially what I did can be replicated as follows:

> If you're using ZFS, make sure dataset on which your Jellyfin app files are located (not the media files themselves) has the same _record size_ as the _page size_ of the `library.db` . (_Note: before running any of these commands make sure your container is stopped, and make a backup of at least the db!_)
```
zfs get recordsize mypool/mydataset
sqlite3 library.db 'pragma page_size;'
```
(_Note: It's also a good idea to make sure the db is not in a corrupt state already:_
```
sqlite3 library.db 'pragma integrity_check;'
```
_If the result of this command is anything other than `ok` then you should not proceed with the rest of these steps._)

As we saw above, the default db page size appears to be `4096`. I would not recommend setting your ZFS dataset's record size this low for a variety of reasons, but I won't go into those here. I have mine set at 64K (and the record size of pretty much any other dataset that has a DB in it.) YMMV. It has worked well for me.

> Instead we can adjust the page size of the `library.db` to match the (64k) record size of the dataset:
```
sqlite3 library.db 'pragma page_size=65536; pragma journal_mode=truncate; VACUUM;'
```

> Since this command sets the journal mode to `truncate`, after the vacuum completes we need to set it back to what it was originally, `delete`:
```
sqlite3 library.db 'pragma journal_mode=delete;
```
(_Note: this is where these steps differ from the ones on the page linked above. The *arrs all use `journal_mode=WAL`. Don't accidentally set the Jellyfin `library.db` to this or there will be problems._)

> Finally, 
```
sqlite3 sonarr.db 'pragma synchronous = Normal;'
```

After doing this, the time my Jellyfin container takes to go from 'start' to when the web UI is available and responsive is < 30s. Previously it was ~5-10 minutes. I just started a full library rescan (~30TB of media) before typing up this post and it's already at 43.8% progress. Clearly there are some risks to doing this, but since I am running ZFS and take regular snapshots, I am not really concerned about incidental DB corruption, since I can always just rollback the snapshot.

I don't think this method will be entirely new to the devs or to anyone who has been following this or other similar issues on the repo, but I haven't seen it typed up anywhere here so I thought I'd do so in case it's useful to anyone. 

I would love to hear people's thoughts on this. Has anyone had success doing something similar?

**Comment by fromSmolsoft** [comment]
It would be nice to add "stripped" scan option as temp workaround to at least discover files without any extra functions.        
Why? In my case the initial / manual scan of 11 videos  (14.4 GB)  in single directory is taking hours and totally freezes any interaction with Jellyfin. (Still not finished, can't tell % because server fails to respond when scan runs).      
At first i though it's slow CPU (usages went to100% ,  error or permissions issue until i found numerous reports of slow scans all over the web.         
I don't wanna to sound ungrateful since it's free soft but it's practically unusable and this issue is at least 4 years old.

**Comment by jocull** [comment]
Disabling real-time monitoring during the initial library scan made a major difference. I don't know if it matters that my `/media` `/config` and `/cache` were all on the same drive during my run? I had a theory that writes to the config or cache would be tripping more disk usage to filter though for monitoring, so disabling the monitor improved the performance. CPU usage was not high during the run, but disk seeking seemed to be very active while disk throughput was low.

**Comment by felix920506** [comment]
@jocull what JF version are you on?

**Comment by jocull** [comment]
@felix920506 I am on `Jellyfin.Server 10.8.13.0` from docker image `ca404505fe6f`

**Comment by sevenrats** [comment]
The vast majority of cases in this thread have no evidence whatsoever that they are bugs.  Any system which scans media from a spinner and then turns around and writes metadata and databases to another spinner (god forbit it's the same one) is going to be insanely slow. That's not a bug thats your hardware going as fast as it can.
@felix920506 i say we close it as not a bug.

**Comment by sevenrats** [comment]
Further, it's imperative that you use a media filesystem which allows for mtimes.  For example, if you mount your media via samba, then you have NO MTIMES and the ENTIRE LIBRARY must be rescanned EVERY TIME. This is also not a bug this is just using a bad filesystem.

**Comment by bigverm23** [comment]
You say that but if you compare the scan times to other platforms like plex or emby....it's significantly longer if it completes at all

**Comment by sevenrats** [comment]
@bigverm23 and I'm not here to say that there was no room for optimization, either, but if you are hairpinning this much data to a spinner, you really have no right to expect a performant UX. if you are using filesystems which don't support mtimes, you are not helping yourself.  The fact is that lot's of optimizations to this effect have already landed in unstable, and imho it would be a shame for this issue to absorb to much of our triage team's time until those changes get released and evaluated by you guys.

**Comment by felix920506** [comment]
There have already been significant improvements in master. Closing this one as completed. Ain't no one gonna read through this thread either way.

Any further findings should be posted as new issue reports. Please be advised that using hardware below [our recommendations](https://jellyfin.org/docs/general/administration/hardware-selection) when reporting performance issues may lead to your report being closed as a non issue.

**Comment by jocull** [comment]
Only here to add additional details since my last post triggered a lot of feelings 🙃

- I’m using directly attached storage over USB.
- Mirrored RAID disks. Yes they are spinning. They aren’t amazing drives, but perfectly functional.
- It’s software RAID under Linux (mdadm) but note again that CPU was not high so I don’t view it as a bottleneck
- It’s using ext4 which seems like a perfectly reasonable file system to choose. Maybe it’s weird when mounting docker volumes?
- It’s a Raspberry Pi 4

The Pi can hash the entire contents of the drives pretty well. I really only thought this was notable because turning off real time monitoring boosted the speed at least 10x

k that’s all love you! ❤️❤️❤️❤️

**Comment by daryl249** [comment]
> Further, it's imperative that you use a media filesystem which allows for mtimes. For example, if you mount your media via samba, then you have NO MTIMES and the ENTIRE LIBRARY must be rescanned EVERY TIME. This is also not a bug this is just using a bad filesystem.

This is the kind of thing that turns off users.  Users that know this is still an issue, users that are running other platforms like plex on the very same machine pointed at the very same libraries on the very same storage and watching plex take 8 minutes to scan with no UI impact and JF take 2 hours and make the UI nearly unusable while only running 60-90% CPU on one core.

It seems that you're just not interested in solving this because presumably it's not a problem for you, and you're obviously upset that people keep reporting it.  How baffling a choice to take a public role in a project like this and care so little about the actual users, blaming them for not reporting the issue properly and/or running the "wrong" hardware.

**Comment by sevenrats** [comment]
@daryl249 like we said, improvements have *already landed* in master and are just waiting to be packaged into release.  Sorry for doing exactly what you asked for I guess.

**Comment by TimGels** [comment]
> It seems that you're just not interested in solving this because presumably it's not a problem for you
If it is this big of a deal to you, why not fix it yourself? We are all volunteers. We welcome pull requests!
